[Lee Yen Shen] iP

leeyenshen opened the pull request:
# IP: Chatbot Tom

## Features Implemented
- Encapsulation and ~~better~~  *much better* **OOP** structure.
- **Event-driven design** for chatbot logic.


## Points of improvement 
1. Enhanced **error handling** with `ChatbotException`.
2. Improved **date parsing** (`LocalDate` & `LocalDateTime` support).
3. Persistent **task storage** using `ChatbotDataHandler`.
> The better chatbot ~Me

## Tasks Completed ✅
- [x] Implement `Event` as an abstract base class.
- [ ] Refactor `Parser` for improved input parsing.
- [ ] Enhance `List` for better task management.
- [x] Refactor `Ui` to improve user interaction.
- [ ] Implement natural language processing for commands (Future Enhancement).

`git status`

## How to Use 🛠️
1. Run the program with:
   ```sh
   java -jar chatbot.jar

## The code behind it
```java
public static void main(String[] args) {
    Tom tom = new Tom("data/Tom.txt");
    new Chatbot(tom).run();
}
🔗 View detailed code changes: [GitHub Repository](https://github.com/leeyenshen/ip)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

leeyenshen committed:
Rename Duke.java to Tom.java

leeyenshen committed:
Initialise skeletal chatbot with greeting and exit functionality

leeyenshen committed:
Add Event.java
Event class is abstract to allow for implementation of Chatbot actions

leeyenshen committed:
Add Chatbot.java Exit.java Greeting.java Listen.java

Added functionality for repeating users command input and exiting when "bye" is typed

leeyenshen committed:
Refactor Tom.java

Refactored main code to use a queue instead to repeat events and implement functionalities.

leeyenshen committed:
Add List.java and functionality for list

Created a list and refactored the code to use a list stored in Listen event to add items and display list based on what the user inputs

leeyenshen committed:
Add Pair.java

Pair.java is the task that stores an item name and boolean of whether its done

leeyenshen committed:
Refactor List.java Listen.java to use Pairs

leeyenshen committed:
Edit List.java marking and unmark methods

Marking and unmark now checks whether it is already done or not done before repeating action

leeyenshen committed:
Edit List.java mark and unmark

List position was not numbered properly

leeyenshen committed:
Add Deadline.java Meeting.java Todo.java

leeyenshen committed:
Refactor Deadline.java Meeting.java Todo.java

Edit code to print out correct output and store it properly

leeyenshen committed:
Refactor List.java, Listen.java

leeyenshen committed:
Add Javadoc comments for all classes and methods

- Documented all classes including Chatbot, Deadline, List, Listen, Meeting, Pair, Todo, Event, Exit, Greeting, and Tom.
- Added detailed descriptions for class purposes, constructors, methods, and return values.
- Improved code readability and maintainability by adhering to Javadoc standards.

leeyenshen committed:
Add automated testing for Tom using I/O redirection

- Implemented a Bash script (`runtest.sh`) to automate compilation, execution, and result verification.
- Added support for input redirection (`input.txt`) and output comparison against expected results (`EXPECTED.TXT`).
- Normalized line endings and whitespace using `dos2unix` and `sed` to ensure consistent comparison.
- Enhanced test script to handle trailing spaces and inconsistent spacing.
- Documented steps for preparing test files and running the script.

leeyenshen committed:
Implement error handling and testing framework for chatbot

- Added comprehensive error handling in the chatbot for invalid commands and input errors.
  - Handled empty descriptions for `todo`, `deadline`, and `event` commands with detailed error messages.
  - Validated proper syntax for `deadline` and `event` commands (e.g., missing `/by`, `/from`, or `/to`).
  - Provided specific error feedback for unknown commands and invalid indices in `mark` and `unmark`.
  - Used `ChatbotException` with dynamic error messages for all cases.
- Enhanced the `Listen` class to process user commands and handle errors gracefully.
- Added a testing framework using I/O redirection:
  - Created a test script (`runtest.sh`) to automate compilation, execution, and comparison of program output.
  - Prepared `input.txt` and `EXPECTED.TXT` for testing typical user interactions.
  - Normalized line endings and whitespace in output for consistent comparison.
- Updated Javadoc comments for all classes and methods:
  - Documented the purpose, parameters, and behavior of each method and class.
  - Included details about exception handling and command processing.

leeyenshen committed:
Add delete functionality for task management

- Implemented the `delete` command in the `List` class to remove tasks by index.
  - Added validation to ensure valid task indices.
  - Displayed confirmation messages after successful deletions.
- Updated the `Listen` class to handle the `delete` command.
  - Integrated the `delete` command into the `processCommand` method.
  - Provided specific error messages for invalid or missing task indices.
- Enhanced user interaction with detailed feedback for task deletions.

This update improves task management by enabling users to remove tasks from the list

leeyenshen committed:
Add comprehensive input and output files for chatbot testing

- Created `input.txt` to test all chatbot functionalities, including:
  - Adding tasks (`todo`, `deadline`, `event`) with valid and invalid inputs.
  - Marking and unmarking tasks, including invalid indices.
  - Deleting tasks, with validation for non-existent and invalid indices.
  - Handling edge cases for commands with missing or incomplete parameters.
  - Proper termination with the `bye` command.
- Added `EXPECTED.TXT` to validate program behavior against:
  - Correct task management operations.
  - Detailed error messages for invalid commands and inputs.
- Ensured the test files cover normal usage scenarios, edge cases, and error handling comprehensively.

This update improves test coverage and ensures the chatbot is robust against various user inputs.

leeyenshen committed:
Refactor: Improve adherence to coding standards in all files

Updated various Java files to follow coding conventions:
- Chatbot.java:
  * Renamed parameter `tom` to `initialEventSource` for clarity.
  * Enhanced Javadoc for the `run` method.

- ChatbotException.java:
  * No changes required; all conventions were followed.

- Deadline.java:
  * Made field `by` private and added a getter method.
  * Optimized `toString` method using `StringBuilder`.

- Event.java:
  * Renamed constant `line` to `LINE` for SCREAMING_SNAKE_CASE compliance.
  * Enhanced class-level Javadoc for clarity.

- Exit.java:
  * Updated Javadoc for `simulate` method to clarify return behavior.

- Greeting.java:
  * Enhanced Javadoc for `simulate` method to describe the returned `Listen` event.

- List.java:
  * Added `final` to method parameters for immutability.
  * Improved Javadoc for constructors and methods.

- Listen.java:
  * Improved class-level Javadoc to describe user interaction.
  * Refactored `processCommand` into smaller helper methods for maintainability.

- Meeting.java:
  * Made fields `from` and `to` private.
  * Optimized `toString` method using `StringBuilder`.

- Pair.java:
  * Made fields `item` and `done` private.
  * Added validation in `mark` method to prevent redundant state changes.

These changes enhance code readability, maintainability, and compliance with Java coding standards.

leeyenshen committed:
Add ChatbotDataHandler.java for task persistence

The chatbot previously lacked task persistence, causing data
to be lost when the application restarted.

Task persistence is crucial for maintaining chatbot state and
ensuring tasks persist across sessions.

Let's:
* Implement `ChatbotDataHandler.java` to manage task storage.
* Create `data/Tom.txt` as the persistent task file.
* Implement `getTasks()` to initialize tasks from memory.
* Implement `saveTasks()` to write tasks back to `data/Tom.txt`.
* Introduce `parseTask()` to process different task types:
  - `T | done | description` → `Todo`
  - `D | done | description | due_date` → `Deadline`
  - `E | done | description | event_from | event_to` → `Meeting`
* Use `LinkedList<Pair>` for efficient task management.
* Print debugging messages when loading and saving tasks.

This enhances chatbot functionality by preserving tasks across
sessions and improving user experience.

leeyenshen committed:
Refactor chatbot files to integrate with ChatbotDataHandler

Previously, the chatbot handled task management without persistent
storage, causing data loss when the application restarted.

Integrating `ChatbotDataHandler` enables persistent task storage,
ensuring tasks are saved and retrieved across sessions.

Let's:
* Modify `Tom.java` to initialize tasks using `ChatbotDataHandler.getTasks()`.
* Update `Exit.java` to call `ChatbotDataHandler.saveTasks()` before exiting.
* Refactor `List.java` to store tasks as `LinkedList<Pair>` for compatibility.
* Modify `Listen.java` to process task commands (`todo`, `deadline`, `event`)
  and update `ChatbotDataHandler`.
* Update `Deadline.java`, `Meeting.java`, and `Todo.java` to include
  `toFileFormat()` for task serialization.
* Improve `Pair.java` to store and format tasks properly.
* Adjust `Greeting.java` to initialize from stored tasks.
* Ensure `Chatbot.java` correctly processes stored task events.

This enhances chatbot functionality by persisting tasks across sessions,
improving usability and reliability.

leeyenshen committed:
Add input validation for task commands

The `mark`, `unmark`, and `delete` commands previously did not
validate user input, causing runtime errors when invalid task
numbers or formats were provided.

Input validation is necessary to prevent errors and ensure the
chatbot provides appropriate feedback for invalid commands.

Let's:
* Add a `try-catch` block for `mark` to handle invalid integers.
* Add a `try-catch` block for `unmark` to handle invalid integers.
* Add a `try-catch` block for `delete` to handle invalid integers.
* Display an error message ("Input a proper number") for invalid inputs.

This improves user experience and ensures robust command handling.

leeyenshen committed:
Update input.txt to add more testcases

leeyenshen committed:
Merge branch 'master' into branch-Level-7

* master:
  Update input.txt to add more testcases

# Conflicts:
#	text-ui-test/input.txt

leeyenshen committed:
Fix spacing in toFileFormat method

The `toFileFormat` method in `Deadline.java` previously had an
incorrect spacing in the output string format for the deadline
task representation.

Proper spacing ensures the file format is consistent and matches
the expected output when tasks are saved and loaded.

Let's:
* Add a space before the pipe (`|`) in the `toFileFormat` method.

This improves consistency and ensures tasks are saved in the
correct format for future processing

leeyenshen committed:
Fix spacing in toFileFormat method

The `toFileFormat` method in `Deadline.java` previously had an
incorrect spacing in the output string format for the deadline
task representation.

Proper spacing ensures the file format is consistent and matches
the expected output when tasks are saved and loaded.

Let's:
* Add a space before the pipe (`|`) in the `toFileFormat` method.

This improves consistency and ensures tasks are saved in the
correct format for future processing

leeyenshen committed:
Update input.txt and expected.txt

Previously, the `input.txt` and its corresponding expected output
did not comprehensively test all chatbot functionalities, such as
edge cases and error handling for various commands.

Testing a wider range of inputs and verifying their output ensures
the chatbot behaves correctly across different scenarios, including
valid, invalid, and edge case inputs.

Let's:
* Update `input.txt` to include comprehensive test cases for:
  - Adding todos, deadlines, and events with valid and invalid inputs.
  - Handling edge cases like empty descriptions and missing fields.
  - Using `mark`, `unmark`, `delete`, and `list` commands.
* Update the expected output to reflect the new test cases.

These changes improve the chatbot's test coverage and ensure its
behavior is robust and predictable.

leeyenshen committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Fix spacing in toFileFormat method
  Fix spacing in toFileFormat method
  Add input validation for task commands
  Refactor chatbot files to integrate with ChatbotDataHandler
  Add ChatbotDataHandler.java for task persistence

# Conflicts:
#	text-ui-test/input.txt

leeyenshen committed:
Refactor task handling and input validation for robustness

Previously, task management in `ChatbotDataHandler.java`, `Deadline.java`,
`Listen.java`, and `Meeting.java` lacked standardized date-time handling,
causing inconsistencies when processing deadlines and events.

Ensuring consistent parsing, formatting, and validation of task inputs
improves reliability and maintains chatbot functionality across various
commands.

Let's:
* Update `ChatbotDataHandler.java` to:
  - Standardize `parseTask()` logic for deadlines and events.
  - Support both `LocalDate` and `LocalDateTime` for flexible input formats.
  - Improve error handling for incorrect date parsing.
* Modify `Deadline.java` to:
  - Differentiate between deadlines with and without time components.
  - Ensure `toFileFormat()` correctly saves deadlines.
* Refactor `Listen.java` to:
  - Improve input validation for `todo`, `deadline`, and `event` commands.
  - Enhance exception handling for missing or malformed inputs.
* Fix `Meeting.java` to:
  - Properly store `LocalDate` and `LocalDateTime` for event scheduling.
  - Ensure correct string representation and file output.

This improves chatbot robustness by ensuring structured date-time
management and better input validation, reducing user input errors.

leeyenshen committed:
Add test case for deadline with time

Previously, there was no test case in `input.txt` to verify if the
chatbot correctly handles deadlines with both date and time specified.

This test is necessary to ensure that the chatbot processes and
saves deadlines with time information accurately, and that output
formatting remains consistent.

Let's:
* Add a test case for the command `deadline say hello /from 2019-02-12 2350`
  in `input.txt`.
* Use this case to validate the correctness of time parsing and output
  formatting in the chatbot's `Deadline` implementation.

This helps verify that the chatbot can properly handle deadlines
that include specific time information and flags potential bugs
in time formatting.

leeyenshen committed:
Merge branch 'Level-8'

* Level-8:
  Refactor task handling and input validation for robustness

leeyenshen committed:
Refactor chatbot to improve OOP structure and modularity

Previously, the chatbot had procedural code with tightly coupled
logic, making it difficult to extend, maintain, and test.

Applying Object-Oriented Programming (OOP) principles enhances
code organization, encapsulation, and separation of concerns.

Let's:
* Introduce `ChatbotDataHandler.java` to handle task persistence.
* Refactor `Tom.java` to initialize `Ui`, `ChatbotDataHandler`,
  and `List` for better modularity.
* Modify `Chatbot.java` to process chatbot events sequentially.
* Extract `ChatbotException.java` for centralized error handling.
* Implement `Event.java` as an abstract class for chatbot events.
* Improve `Listen.java`, `Greeting.java`, and `Exit.java` to
  handle chatbot states using event-driven design.
* Enhance `List.java` to encapsulate task operations (`add`, `mark`,
  `unmark`, `delete`) independently.
* Improve `Deadline.java` and `Meeting.java` to support both `LocalDate`
  and `LocalDateTime` formats for better date-time parsing.
* Add `Parser.java` to handle flexible date/time parsing with
  improved error messages.

This refactor improves maintainability, scalability, and readability,
allowing for easier feature expansion and better testability.

leeyenshen committed:
Expand input.txt for comprehensive chatbot testing

Previously, `input.txt` lacked sufficient test coverage for edge cases,
error handling, and various chatbot functionalities.

Ensuring a wider range of test cases improves the chatbot's robustness
by validating its behavior across different inputs, including valid,
invalid, and boundary cases.

Let's:
* Add tests for `todo`, `deadline`, and `event` commands with both
  valid and invalid inputs.
* Include boundary cases such as empty descriptions and missing fields.
* Test `mark`, `unmark`, and `delete` commands with valid and out-of-bounds indices.
* Verify chatbot persistence by adding tasks, exiting, and restarting.
* Ensure error messages appear correctly for invalid operations.

These additions enhance chatbot test coverage, ensuring reliable
handling of various user inputs.

leeyenshen committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Refactor chatbot to improve OOP structure and modularity

leeyenshen committed:
Move all classes into a single package for better organization

Previously, classes in `src/main/java` were unstructured, making it
harder to manage dependencies and maintain consistency.

Using a single package improves project organization, enhances
modularity, and simplifies imports.

Let's:
* Move all classes into `com.example.project` (replace with actual package).
* Update `package` declarations in all Java files.
* Adjust import statements to match the new package structure.

This change ensures a cleaner, more maintainable codebase by
consolidating all classes under a well-defined package structure.

leeyenshen committed:
Add comprehensive tests for chatbot error handling

Previously, `input.txt` lacked extensive test coverage for edge
cases, invalid inputs, and unexpected user commands.

Ensuring robust error handling improves chatbot reliability by
validating its response to incorrect, incomplete, or malformed
inputs.

Let's:
* Add tests for invalid `todo`, `deadline`, and `event` commands:
  - Missing descriptions and required fields.
  - Incorrect date and time formats.
  - Impossible dates and negative values.
* Add boundary tests for `mark`, `unmark`, and `delete`:
  - Non-existent task indices, negative numbers, and zero.
  - Non-numeric inputs and extra arguments.
* Include invalid and mistyped commands to test chatbot robustness.
* Verify persistence by adding, listing, and deleting tasks
  across chatbot sessions.

These test cases enhance chatbot error handling, ensuring it
responds correctly to all invalid inputs.

leeyenshen committed:
Merge branch 'A-Gradle'

* A-Gradle:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

leeyenshen committed:
Refactor delete method to streamline output handling

- Refactor the `delete` method to remove the intermediate `output` variable.
- Directly return the formatted string response, reducing redundancy and improving clarity.
- Update the exception message to be more descriptive and user-friendly.

This change enhances code readability and simplifies the logic in the `delete` method while ensuring consistency across similar methods.

leeyenshen committed:
Set up Gradle to run Tom

Previously, the Gradle build configuration was incomplete and did
not allow the `Tom` application to run seamlessly.

Setting up Gradle to execute `Tom` ensures a smoother development
workflow and allows easy building, running, and packaging of the
application.

Let's:
* Update `build.gradle`:
  - Set `application.mainClass` to `tom.Tom` to define the main
    entry point for the application.
  - Configure `shadowJar` to generate a JAR named `Tom.jar`.
  - Add a `run` task to accept standard input during execution.

These changes enable the project to:
1. Be run directly using `./gradlew run`.
2. Generate a standalone JAR file with `./gradlew shadowJar`.

This setup improves developer productivity and ensures consistent
execution across environments.

leeyenshen committed:
Update Gradle configuration for Tom project

Previously, the Gradle build script referenced the old project name
"Duke" and lacked clarity in its structure.

Updating the configuration ensures the build process aligns with
the new project name "Tom" and improves readability.

Let's:
* Update `application.mainClass` to `tom.Tom` to reflect the main
  entry point of the application.
* Rename the shadow JAR output to `Tom.jar` for consistency with
  the new project name.
* Fix minor formatting in the Gradle `run` task for better readability.

These changes improve the build process, ensuring the project can
be built and executed correctly as "Tom."

leeyenshen committed:
add JUnit tests for Parser and ChatbotDataHandler

Add JUnit tests to ensure robust behaviour of key methods in Parser and
ChatbotDataHandler.

- Add ParserTest.java:
  - Test valid commands (`todo`, `deadline`, `bye`).
  - Verify invalid commands print correct error messages.
  - Ensure correct task additions in the list.

- Add ChatbotDataHandlerTest.java:
  - Test `saveTasks()` with an empty list and multiple tasks.
  - Verify correct file format and ensure data persistence.
  - Check if the file is created when missing.

These tests improve code reliability by covering expected functionality
and potential edge cases.

leeyenshen committed:
Add Javadocs to ChatbotDataHandler, Parser, and Ui classes

- Documented all public methods and constructors with detailed Javadocs.
- Improved code readability and ensured compliance with SE-EDU Java Coding Standards.
- Included usage of {@code} tags and structured descriptions for clarity.

leeyenshen committed:
Apply SE-EDU coding standard fixes to multiple files

- Standardized variable names (`timeby` → `timeBy`, `dateby` → `dateBy`).
- Updated Javadoc descriptions for clarity in Event, Greeting, and Listen.
- Changed `FILE_PATH` to `private final` in ChatbotDataHandler.
- Fixed incorrect event type in Meeting's `toFileFormat()`.
- Improved error messages in List class for professionalism.

leeyenshen committed:
Apply SE-EDU coding standard fixes to multiple files

- Updated exception messages in Parser to be more professional.
- Corrected Javadoc references in Tom and Pair.
- Ensured method documentation clarity across all files.

leeyenshen committed:
Add find command to search tasks by keyword

- Added `find` method in List to search tasks containing a keyword.
- Updated Parser to handle `find` command.
- Improved Ui formatting for search results.

leeyenshen committed:
Merge branch 'Level-9'

* Level-9:
  Add find command to search tasks by keyword

leeyenshen committed:
Merge branch 'A-JavaDoc'

* A-JavaDoc:
  Add Javadocs to ChatbotDataHandler, Parser, and Ui classes

Noob-No-1 commented on src/main/java/tom/ChatbotDataHandler.java:
```diff
@@ -0,0 +1,176 @@
+import java.util.LinkedList;
+import java.io.FileWriter;
+
+/**
+ * Handles the loading and saving of chatbot task data from a file.
+ */
```
Great job on not forgetting a documentation for the entire class! (I did)

Noob-No-1 commented on src/main/java/tom/ChatbotDataHandler.java:
```diff
@@ -0,0 +1,176 @@
+
+    public ChatbotDataHandler(String filePath) {
+        this.FILE_PATH = filePath;
+    }
+    public List getTasks() throws ChatbotException, IOException {
```
Maybe it will look cleaner if you can separate two methods by leaving an empty line between them?

Noob-No-1 commented on src/main/java/tom/ChatbotDataHandler.java:
```diff
@@ -0,0 +1,176 @@
+    /**
+     * Parses a task string and creates the appropriate Task object.
+     * Format:
+     * - tom.Todo: "T | <done> | <description>"
+     * - tom.Deadline: "D | <done> | <description> | <due_date>"
+     * - tom.Event: "E | <done> | <description> | <event_from> | <event_to>"
```
I like the specification of what output will look like in the documentation!

Noob-No-1 commented on src/main/java/tom/Deadline.java:
```diff
@@ -0,0 +1,64 @@
+ * Represents a task with a deadline.
+ * A tom.Deadline includes a description of the task and a date/time by which it must be completed.
+ */
+public class Deadline extends Pair{
```
Very interesting implementation, but maybe name the Pair class differently or change the association between this class and pair to be composition? So as to avoid confusion due to naming issue.

Noob-No-1 commented on src/main/java/tom/Event.java:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * A separator line for output formatting.
+     */
+    public static String LINE = "========================================";
```
Maybe it makes more sense to put it in the Main class or the UI class? Since the separate line is not a unique attribute to event class but rather UI in general.

Noob-No-1 commented on src/main/java/tom/Ui.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public void startUp() {
+        String logo = "  _______   ____    __  __ \n"
+                    + " |__   __| /  _ \\  |  \\/  |\n"
+                    + "    | |    | | | | | |\\/| |\n"
+                    + "    | |    | |_| | | |  | |\n"
+                    + "    |_|    \\____/  |_|  |_|\n";
```
Nice! :)

Noob-No-1 commented on src/main/java/tom/Event.java:
```diff
@@ -0,0 +1,22 @@
+ * Subclasses should implement the `simulate` method to define
+ * specific event behaviors.
+ */
+public abstract class Event{
```
Maybe change the naming of this class to Task? The naming is a bit confusing as it coincides with a non-abstract object event as required by the course ip website.

leeyenshen committed:
Integrate JavaFX UI with chatbot

The chatbot previously ran entirely in the console and lacked a graphical
interface. Additionally, chatbot messages had no structured display,
and resizing the window did not adjust UI elements properly.

This commit integrates JavaFX to provide a GUI for the chatbot.

- **Add JavaFX UI Components**:
  - Implement `MainWindow.fxml` for the main chatbot interface.
  - Implement `DialogBox.fxml` to display user and chatbot messages.
  - Add `MainWindow.java` as the controller to handle user input.
  - Add `DialogBox.java` to style chat messages with images and borders.

- **Ensure UI Responsiveness**:
  - Use `AnchorPane` constraints to make `MainWindow.fxml` resize dynamically.
  - Ensure `ScrollPane`, `VBox`, `TextField`, and `Button` expand properly.

- **Improve Chat Message Display**:
  - Add a black-bordered text box for chatbot messages (`DialogBox.fxml`).
  - Use rounded corners and padding for readability.

- **Redirect Chatbot I/O to GUI**:
  - Modify `Tom.java` to return responses instead of printing them.
  - Update `Ui.java` to return strings rather than using `System.out.println()`.
  - Modify `MainWindow.java` to call `tom.getResponse(input)`,
    ensuring user input is processed and displayed in the UI.

With these changes, the chatbot now runs in a **fully interactive JavaFX GUI**,
making it visually structured, responsive, and user-friendly.

leeyenshen committed:
Merge branch 'Level-10'

* Level-10:
  Integrate JavaFX UI with chatbot

hrishikeshsathyian commented on src/main/java/tom/ChatbotDataHandler.java:
```diff
@@ -0,0 +1,176 @@
+ * Handles the loading and saving of chatbot task data from a file.
+ */
+public class ChatbotDataHandler {
+    private String FILE_PATH;
```
Should we be using a different casing here? That casing is meant for final constants, which is not reflected in the code line

hrishikeshsathyian commented on src/main/java/tom/Exit.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public Event simulate() {
+        this.chatbotDataHandler.saveTasks(this.list);
+        new Ui().exit();
```
Is there any reason for creating a new Ui for each event subclass, instead of one main instance of Ui that is passed down to all the Event objects?

hrishikeshsathyian commented on src/main/java/tom/Parser.java:
```diff
@@ -0,0 +1,167 @@
+                    return list.add(new Meeting(eventParts[0].trim(), false, fromDate, toDate));
+                }
+            } catch (DateTimeParseException e) {
+                throw new ChatbotException("Invalid date format: " + from + to +
```
Should the + operator be on the new line instead

hrishikeshsathyian commented on src/main/java/tom/List.java:
```diff
@@ -0,0 +1,129 @@
+     */
+    public String find(String keyword) {
+        StringBuilder result = new StringBuilder("Here are the matching tasks in your list:\n");
+        boolean found = false;
```
```suggestion
        boolean isFound = false;
```

Should this be named isFound instead?

hrishikeshsathyian reviewed (commented):
Overall, looks very good, just some small naming convention errors and minor nits. Wonderful Code!

leeyenshen committed:
Add addInitialBotMessage() for chatbot initialisation

Currently, the chatbot's initial response is not modularised, making
initialisation handling less structured.

Let's add a new method `addInitialBotMessage()` to:
- Display the chatbot's initial message when `setTom()` is called.
- Prevent `NullPointerException` by ensuring `tom` is not null before use.
- Improve code readability and maintainability.

This ensures a cleaner and more structured chatbot initialisation process.

leeyenshen committed:
Add assertions for validation

Assertions are currently not used in the codebase, which may lead to
unexpected null values, incorrect indices, and invalid states passing
silently during execution.

Adding assertions in key areas of `Pair`, `Deadline`, `Meeting`, `List`,
and `Ui` ensures that:
- Task descriptions and deadlines are non-null upon object creation.
- Meeting start dates/times are before end dates/times.
- Task indices in `mark` and `delete` operations are within valid bounds.
- The scanner in `Ui` is available before reading input.

This improves code reliability by catching potential logic errors
early during execution.

leeyenshen committed:
Refactor deep nesting and improve clarity

Previously, several methods contained deeply nested logic, making them
difficult to read and maintain. In particular:
- `Deadline.toString()` had redundant conditionals.
- `Exit.simulate()` buried logic inside unnecessary nesting.

Refactored to:
- Reduce nesting in `Deadline.toString()` by using early returns.
- Make the happy path prominent in `Exit.simulate()`.

This improves code readability and maintainability.

leeyenshen committed:
Replace magic numbers

Magic numbers and inconsistent naming conventions were found in:
- `List.mark()` using `0` instead of a named constant.

Changes:
- Introduced `MIN_INDEX` constant in `List.java` for clarity.

This improves maintainability and avoids confusion in variable meanings.

leeyenshen committed:
Extract duplicate logic and improve error handling

Previously:
- Catch block in `ChatbotDataHandler` silently ignored errors.

Changes:
- Improved exception handling by logging errors instead of ignoring them.

This enhances maintainability and debugging support.

leeyenshen committed:
Merge pull request #1 from leeyenshen/A-Assertions

Add assertions for validation

leeyenshen committed:
Merge commit '8a2ea000c6f8cb5ff901ca07a1bfe6b027af1201' into A-CodeQuality

* commit '8a2ea000c6f8cb5ff901ca07a1bfe6b027af1201':
  Add assertions for validation

# Conflicts:
#	src/main/java/tom/List.java

leeyenshen committed:
Prevent duplicate tasks from being added

Previously, the chatbot allowed users to add duplicate tasks with
the same description, leading to redundant entries in the task list.

To fix this:
- Introduced a `containsDuplicate()` method in `List.java` to check for duplicate tasks.
- Ensured that duplicate tasks are not added, improving data integrity.

This update prevents redundancy and enhances task list management.

leeyenshen committed:
Merge branch 'BCD-Extension'

* BCD-Extension:
  Prevent duplicate tasks from being added

leeyenshen committed:
Add files via upload

leeyenshen committed:
Refactor Ui.png documentation to docs

Refactor Ui.png to docs by converting relevant documentation into a structured format for better maintainability and readability.

- Ensured content is structured for clarity.
- Improved documentation consistency with other project files.
- Facilitated easier updates and modifications.

This refactor enhances the documentation's organization, making it more accessible for future reference and updates.

leeyenshen committed:
Merge commit 'c19b6edc708eae135046f1fd18855e197459eb75'

* commit 'c19b6edc708eae135046f1fd18855e197459eb75':
  Extract duplicate logic and improve error handling
  Replace magic numbers
  Refactor deep nesting and improve clarity

leeyenshen committed:
add README.md to docs directory

The project currently lacks a README.md file in the docs directory,
which is essential for providing documentation and usage guidelines.

Adding docs/README.md helps in organizing project documentation
effectively and ensures that developers can find relevant information
easily.

This commit introduces docs/README.md as a starting point for
documentation.

leeyenshen committed:
update README.md with usage instructions

The project currently lacks a README.md file in the docs directory,
which is essential for providing documentation, usage instructions,
and setup guidelines.

Adding docs/README.md improves developer and user accessibility by
offering:
- A clear overview of the chatbot's functionality.
- Step-by-step instructions on how to set up and run the chatbot.
- Example commands to interact with the chatbot.
- Guidance on how tasks such as adding, deleting, and marking tasks
  as complete can be performed.

This commit introduces docs/README.md as a starting point for
documentation, making it easier for users to understand and use the
product.

leeyenshen committed:
Add Javadoc comments for event and task operations

The `Parser` class lacks Javadoc comments for several key methods,
making it difficult to understand their purpose and expected behavior.

Adding Javadoc comments to `parseEvent`, `processMark`, `processUnmark`,
and `processDelete` improves code readability and maintainability.
These comments adhere to the SE-EDU Java coding standard.

With this change:
- `parseEvent` now has documentation explaining event parsing and
  validation.
- `processMark` and `processUnmark` have descriptions clarifying
  how tasks are marked as done/undone.
- `processDelete` is now documented to specify how task removal
  is handled and potential exceptions.

This improves overall code documentation and ensures adherence to
SE-EDU Java best practices.

leeyenshen committed:
refactor boolean variable naming for clarity

Boolean variables in `Pair.java` and `List.java` are named ambiguously,
reducing code readability and making their purpose less clear.

Renaming:
- `done` → `isDone` in `Pair.java` to clearly indicate task completion status.
- `found` → `isFound` in `List.java` to explicitly denote if a match was found.

These changes enhance readability and maintainability by following
SE-EDU's boolean naming convention, which recommends using `is`/`has`
prefixes for boolean values.

leeyenshen committed:
refactor parseTask() into smaller methods

The parseTask() method in ChatbotDataHandler is currently too large
and handles multiple responsibilities, making it difficult to read
and maintain.

Breaking it down into smaller methods will improve readability,
modularity, and error handling.

Let's refactor parseTask() by:
- Extracting parseTodo(), parseDeadline(), and parseEvent() for
  parsing different task types.
- Creating parseDeadlineDate() and parseEventDates() to handle
  date parsing separately.
- Adding error handling for invalid date formats.

This improves code maintainability, making future modifications easier.

leeyenshen committed:
Improve Javadocs for Tom.main()

The Javadocs for the main() method are unclear and do not fully
describe its purpose.

Updating the Javadocs improves clarity by:
- Stating that main() is the entry point of the chatbot application.
- Describing its role in initializing and starting the chatbot.
- Using concise and professional phrasing aligned
  with SE-EDU conventions.

This improves code documentation, making it easier for developers
to understand the purpose of main().
